### ISSUE TITLE
[RFC] Support env-dependent config in one yaml file

### ISSUE THREAD

**Issue by nicolas-grekas** [symfony/symfony#40215]
Inspired by https://github.com/symfony/symfony/pull/40214

For routing/DI, it might be worth merging per-env yaml files (as found in recipes) in one, don't you think?

My proposal would be to go with a suffix:

```yaml
services@dev:
    App\FooForDev: ~

framework:
    secret: '%env(APP_SECRET)%'

framework@test:
    test: true
    session:
        storage_factory_id: session.storage.mock_file
```

WDYT?

**Comment by nicolas-grekas** [comment]
Alternative, using yaml tags:
```yaml
prod: !env
    services:
        App\FooForDev: ~

framework:
    secret: '%env(APP_SECRET)%'

test: !env
    framework:
        test: true
        session:
            storage_factory_id: session.storage.mock_file
```

**Comment by kaznovac** [comment]
imho this is not worth the implementation effort - it's somewhat burdensome (as it requires developer to learn extra micro-syntax), and it's not substantially better than the folder solution already in place.

here are my 2 penny worth on the somewhat more flexible but more verbose yaml syntax:
```yaml
services:
    App\FooForDev:
        class: ~
        condition: '@=container.getEnv('APP_ENV') == "test"' # symfony expression

framework:
    secret: '%env(APP_SECRET)%'

conditional:
    condition: '@=container.getEnv('APP_ENV') == "test"' 
        framework:
            test: true
            session:
                storage_factory_id: session.storage.mock_file
```


in the long run i'd like to see symfony strive to use **php configuration** more (and phase out yaml/xml) - it's simply more ide friendly and easier to parse in run/compile time, and i personally find it easier to understand (it's more concise and has deterministic meaning) and with less cognitive load.

**Comment by javiereguiluz** [comment]
Here's a proposal for an alternative notation: use `@dev`, `@prod`, `@test` as the special prefixes under which you can define services/params only for dev/prod/test:

```yaml
framework:
    secret: '%env(APP_SECRET)%'

@dev:
    services:
        App\FooForDev: ~

@test:
    framework:
        test: true
        session:
            storage_factory_id: session.storage.mock_file
```

**Comment by nicolas-grekas** [comment]
Thank you for the proposal @javiereguiluz, I love it!

I now implemented this proposal for all config formats in https://github.com/symfony/symfony/pull/40214.

Writing this implementation is what made me decide between Javier's proposal and mine using suffixes (see description above): using one extra nesting level works well in all formats.

**Comment by nicolas-grekas** [comment]
@kaznovac thanks for sharing your opinion on this. About `container.getEnv('APP_ENV')`, this is exactly what should be avoided: hardcoding some specific way to access the env, while there is already a more generic mechanism. Actually, this is a great example of the risk with PHP format: ppl can write whatever logic they think about, and just get it wrong...

**Comment by nicolas-grekas** [comment]
@wouterj spotted a significant issue with your proposal Javier: it's not legal in Yaml to start a value with an `@`, unless the string is quoted.`'@test':` might be less nice. PR updated to use `when@dev` instead.

~I'm going to try the suffix approach also, to see how that impacts the implementation.~ (actually not, because the suffix doesn't work for routes)

**Comment by gggeek** [comment]
Although I see potential for this, I am also quite loath to add even _more_ flexibility to the configuration system.
As it stands, when I set up any modern, somewhat complex app based on Symfony, I often find myself deeply lost in the configuration.

- there's a plethora of config files, in /config and in /vendor bundles
- the order in which they are loaded is not clear - let alone there's not even a standard for naming or ordering config files
- config can be made up of parameters, services and semantic config. 3 different building blocks
- parameters are generally used to set values into services, but sometimes compilerPasses and whatnot add or modify parameters
- same with service definitions
- any of the above can actually come from an env var
- which might be overridden by one of many .env files
- and modified via an env processor
- did I mention flex recipes and how they can kick in also at any time to modify the above, with code which is not even stored in the vendor's source code, but in an external repository?

Don't get me wrong, I like flexibility, layered architectures, automation and BC as much as the next guy, but I start to wonder where we will have to draw a line...

**Comment by nicolas-grekas** [comment]
I agree with this @gggeek, and that's precisely why I opened this PR: it will allow reducing the number of files to open to get the details of the configuration settings.

**Comment by jkufner** [comment]
Just drop YAML and use PHP.

**Comment by nicolas-grekas** [comment]
@jkufner that's a common misconception: php doesn't have access to the env right now. This PR enables it.

**Comment by jkufner** [comment]
Main benefit of configuration files is simplicity and declarative nature. The more complicated the syntax and more powerful features, the more difficult it is to use and debug. A few more iterations and we have a programming language which is too complicated, ugly, unsupported by IDE, poorly documented, formally messed up, full of compatibility compromises, … please don't do it.

I really like the per-env directories. It is simple and obvious. I also like `.env.*.local` and `.env.*` files. It is exaclty what is mostly needed. If you need more powerful configuration, you are likely overengineering it.

Not so long ago I've read something about promoting PHP config in Symfony. (I think something in relation to https://www.reddit.com/r/PHP/comments/h93fjc/symfony_will_now_promote_php_over_yaml_and_xml/.) We should go there.

**Comment by jkufner** [comment]
> @jkufner that's a common misconception: php doesn't have access to the env right now. This PR enables it.

Not having access to the env is a good thing. Less moving parts the better.

**Comment by nicolas-grekas** [comment]
Note that in case you missed it, this is already implemented in #40214.
I think it's a good addition to the config capabilities or Symfony.
Varying by env is extremely common, as that's the very purpose of the concept of env.
